lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <mutex>
+#include <optional>
+#include <thread>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/TracingCommon.h"
+#include "revng/Support/Assert.h"
+
+inline constexpr auto TracingEnv = "REVNG_C_API_TRACE_PATH";
+static bool TracingEnabled = llvm::sys::Process::GetEnv(TracingEnv).hasValue();
+
+// The opposite of a std::recursive_mutex, if locked by the same thread it will
+// assert (this is to avoid a deadlock/malformed output when tracing)
+class NonRecursiveMutex {
+private:
+  std::optional<std::thread::id> ThreadId;
+  std::mutex TheLock;
+
+public:
+  NonRecursiveMutex() {}
+
+  void lock() {
+    if (ThreadId.has_value())
+      revng_assert(std::this_thread::get_id() != *ThreadId,
+                   "NonRecursiveMutex entered twice by the same thread!");
+    TheLock.lock();
+    ThreadId = std::this_thread::get_id();
+  }
+
+  void unlock() {
+    revng_assert(std::this_thread::get_id() == *ThreadId);
+    TheLock.unlock();
+    ThreadId.reset();
+  }
+};
+
+static NonRecursiveMutex TraceLock;
+
+// Helper class for tracing, this will be used by the the argument/return
+// handlers defined below to properly print the value of the arguments onto the
+// YAML tracing file
+class PipelineCTracer {
+private:
+  std::string Path;
+  std::error_code OpenEC;
+  llvm::raw_fd_ostream OS;
+  bool OutputtingArguments;
+
+public:
+  PipelineCTracer(std::string Path) :
+    Path(Path), OpenEC(), OS(Path, OpenEC), OutputtingArguments(false) {
+    revng_assert(!OpenEC);
+    printHeader();
+  }
+
+  ~PipelineCTracer() {
+    OS.flush();
+    OS.close();
+  }
+
+private:
+  void printHeader() {
+    OS << "version: 1\n";
+    OS << "commands:\n";
+    OS.flush();
+  }
+
+  std::string reprString(const char *String) {
+    return '"' + llvm::StringRef(String).str() + '"';
+  }
+
+public:
+  void functionPrelude(const llvm::StringRef Name) {
+    OS << "- name: " << Name << "\n";
+    OS << "  arguments:\n";
+    OutputtingArguments = true;
+    OS.flush();
+  }
+
+  void newArgument() {
+    OS << "  - ";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires std::is_same_v<T, uint8_t> ||
+           std::is_same_v<T, uint32_t> ||
+           std::is_same_v<T, uint64_t>
```

Comment:
```suggestion
template<integral T>
```

Reply:
`integral` includes `char`s which I'd really like to avoid

Reply:
OK, but if you have a require expression used more than once define a concept.

Reply:
I've outlined it in `isInteger`

> What about the poor uint16_t?

With the exception of `rp_initialize` (which we could change) we only really use `int` and `uint64_t`

Reply:
OK, but I'd try to think this as component with its own dignity that does a specific thing that can be described in a general way.

Think this as a feature that others might reuse if they choose a similar approach to our C API (no `struct`, just opaque pointers...).
What would they want to happen if they have a `uint16_t`?

Reply:
Ok, but it still leads to a half-way situation.
The tracing code makes some (not many) assumptions on how things work in PipelineC (e.g. `_destroy` requiring pointer arguments, RPType being "special"). All of these could be "abstracted" away, but I don't see the advantage in doing it, I'd rather KISS it and only adopt the data types we actually use in PipelineC. If we ever get to the point where it's beneficial to recycle this logic then we'll pay for the abstraction part

Reply:
Reworked a bit, added `uint16_t`
Tracing should be as reusable as possible provided a few definitions (e.g. `isRPType`)